src/SDPoptimize.jl

```diff
@@ -238,168 +142,206 @@ function sdp_optimize(model::SPModel,
-    V = zeros(Float64, param.totalStateSpaceSize, TF+1)
-    x = zeros(Float64, model.dimStates)
-    x1 = zeros(Float64, model.dimStates)
+    next_state = zeros(Float64, model.dimStates)
     law = model.noises
 
-    u_lower_bounds = [ i for (i , j) in model.ulim]
-    x_lower_bounds = [ i for (i , j) in model.xlim]
+    u_bounds = model.ulim
+    x_bounds = model.xlim
+    x_steps = param.stateSteps
+
+    #Compute cartesian product spaces
+    product_states, product_controls = generate_grid(model, param)
 
-    count_iteration = 1
+    V = zeros(Float64, param.stateVariablesSizes..., TF)
 
     #Compute final value functions
-    for indx = 1:(param.totalStateSpaceSize)
-        x = variable_from_index(indx, x_lower_bounds, param.stateVariablesSizes,
-                                param.stateSteps)
-        V[indx, TF+1] = model.finalCostFunction(x)
+    for x in product_states
+        ind_x = index_from_variable(x, x_bounds, x_steps)
+        V[ind_x..., TF] = model.finalCostFunction(x)
     end
 
     #Construct a progress meter
-    p = Progress(TF*param.totalStateSpaceSize, 1)
+    p = Progress((TF-1)*param.totalStateSpaceSize, 1)
 
     #Display start of the algorithm in DH and HD cases
-    #Define the loops order in sdp
     if (param.infoStructure == "DH")
         if display
-            println("Starting stochastic dynamic programming
-                    decision hazard computation")
+            println("Starting stochastic dynamic programming decision hazard computation")
         end
 
         #Loop over time
-        for t = TF:-1:1
-            count_iteration = count_iteration + 1
+        for t = (TF-1):-1:1
+            Vitp = value_function_interpolation(model, V, t+1)
 
             #Loop over states
-            for indx = 1:(param.totalStateSpaceSize)
+            for x in product_states
 
                 if display
                     next!(p)
                 end
 
-                v = Inf
-                v1 = 0
-                indu1 = -1
-                Lv = 0
-                x = variable_from_index(indx, x_lower_bounds,
-                                        param.stateVariablesSizes,
-                                        param.stateSteps)
+                expected_V = Inf
+                optimal_u = tuple()
+                current_cost = 0
 
                 #Loop over controls
-                for indu = 1:(param.totalControlSpaceSize)
+                for u = product_controls
 
-                    v1 = 0
-                    count = 0
-                    u = variable_from_index(indu, u_lower_bounds,
-                                            param.controlVariablesSizes,
-                                            param.controlSteps)
+                    expected_V_u = 0
+                    count_admissible_w = 0
 
                     #Loop over uncertainty samples
-                    for w = 1:param.monteCarloSize
+                    if (param.expectation_computation=="MonteCarlo")
```

Comment:
I didn't create an auxialiary function but tuned the expectation computation parameters before the loop. Is it ok?